File: src/Alimer/Engine/Components/LightComponent.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using Alimer.Rendering;

namespace Alimer.Engine;

/// <summary>
/// A component that contains a mesh and materials for the visual appearance of an entity.
/// </summary>
[Meta]
[DefaultEntitySystem(typeof(LightSystem))]
public sealed partial class LightComponent : Component
{
    public LightComponent()
    {
    }

    public LightType LightType { get; set; } = LightType.Point;
    public Color Color { get; set; } = Colors.White;
    public float Intensity { get; set; } = 1.0f;
}

File: Source/Alimer/Renderer/Components/LightComponent.cpp
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "Alimer/Renderer/Components/LightComponent.h"

using namespace Alimer;

void LightComponent::Register()
{
    auto reflection = GetTypeInfoReflection(LightComponent::GetTypeInfoStatic());
    reflection->SetFactory<LightComponent>();
    reflection->SetCategory("Rendering");
    reflection->SetDisplayName("Light Component");

    RegisterEnumProperty("type", &LightComponent::GetLightType, &LightComponent::SetLightType, kDefaultLightType);
    RegisterRefProperty("color", &LightComponent::GetColor, &LightComponent::SetColor, kDefaultLightColor);
    RegisterProperty("intensity", &LightComponent::GetIntensity, &LightComponent::SetIntensity, kDefaultLightIntensity);
    RegisterProperty("range", &LightComponent::GetRange, &LightComponent::SetRange, kDefaultLightRange);
}

LightComponent::LightComponent(LightType type)
    : _lightType(type)
{
    _requireTransformChangeListener = true;
}

void LightComponent::SetLightType(LightType value)
{
    if (_lightType == value)
        return;

    _lightType = value;
}

void LightComponent::SetColor(const Color& value)
{
    if (_color == value)
        return;

    _color = value;
}

void LightComponent::SetIntensity(float value)
{
    if (_intensity == value)
        return;

    _intensity = value;
}

void LightComponent::SetRange(float value)
{
    if (_range == value)
        return;

    _range = value;
}
